app/src/main/java/com/github/onlynotesswent/model/folder/FolderRepositoryFirestore.kt

```diff
@@ -199,39 +350,120 @@ class FolderRepositoryFirestore(private val db: FirebaseFirestore) : FolderRepos
+                    onSuccess = {},
+                    onFailure = {
+                      onFailure(it)
+                      Log.e(TAG, "Failed to delete folder contents: ${it.message}")
+                    },
+                    useCache = true)
+                noteDao.deleteNotesFromFolder(subFolder.id)
+                folderDao.deleteFolderById(subFolder.id)
+              }
+            }
+            onSuccess()
+          },
+          onFailure = { e: Exception ->
+            Log.e(TAG, "Failed to delete folder contents: ${e.message}")
+            onFailure(e)
+          },
+          useCache = true)
+    }
+
     getSubFoldersOf(
-        folder.id,
+        parentFolderId = folder.id,
         onSuccess = { subFolders ->
           subFolders.forEach { subFolder ->
-            deleteFolderContents(
-                folder = subFolder,
-                noteViewModel = noteViewModel,
-                onSuccess = {},
-                onFailure = {
-                  onFailure(it)
-                  Log.e(TAG, "Failed to delete folder contents: ${it.message}")
-                })
-            noteViewModel.deleteNotesFromFolder(subFolder.id)
-            deleteFolderById(
-                folderId = subFolder.id,
-                onSuccess = {},
-                onFailure = {
-                  onFailure(it)
-                  Log.e(TAG, "Failed to delete folderContents: ${it.message}")
-                })
+            CoroutineScope(Dispatchers.IO).launch {
+              deleteFolderContents(
+                  folder = subFolder,
+                  noteViewModel = noteViewModel,
+                  onSuccess = {},
+                  onFailure = {
+                    onFailure(it)
+                    Log.e(TAG, "Failed to delete folder contents: ${it.message}")
+                  },
+                  useCache = useCache)
+              noteViewModel.deleteNotesFromFolder(subFolder.id)
+              deleteFolderById(
+                  folderId = subFolder.id,
+                  onSuccess = {},
+                  onFailure = {
+                    onFailure(it)
+                    Log.e(TAG, "Failed to delete folderContents: ${it.message}")
+                  },
+                  useCache = useCache)
+            }
           }
           onSuccess()
         },
         onFailure = { e: Exception ->
-          onFailure(e)
           Log.e(TAG, "Failed to delete folder contents: ${e.message}")
-        })
+          onFailure(e)
+        },
+        useCache = useCache)
+  }
+
+  /**
+   * Synchronizes the Firestore database with the local cache based on `lastModified` field. Should
+   * be called only if the userId of the folder is the current user.
+   *
+   * @param firestoreFolder The folder from Firestore.
+   * @param cachedFolder The folder from the local cache.
+   * @return Latest version of the folder.
+   */
+  private suspend fun syncFolderFirestoreWithCache(
```

Comment:
Why not just cache the folder if necessary, instead of calling `addFolder` which caches it + updates firebase (not useful, since the cache is never going to be more updated than the firestore file as we have a write-through cache system: whenever you update a folder you update it in firestore)
This way we avoid a redundant call to the repository

Reply:
In the situation where we update the notes (or folders in this case) and there is no internet connection available, the update will be seen only in the cache and not in firestore. By updating both, we can assure that firestore and cache are up to date with the same version of the data.

Reply:
Good point!